toolchain/base/value_store.h

```diff
@@ -113,9 +175,48 @@ class ValueStore
   }
 
  private:
+  // Poison the entire contents of the value store. This is used to detect cases
+  // where references to elements in a value store are used across calls that
+  // might modify the store.
+  auto PoisonAll() const -> void {
+#if CARBON_POISON_VALUE_STORES
+    if (!all_poisoned_) {
+      __asan_poison_memory_region(values_.data(),
+                                  values_.size() * sizeof(values_[0]));
+      all_poisoned_ = true;
+    }
+#endif
+  }
+  // Unpoison the entire contents of the value store.
+  auto UnpoisonAll() const -> void {
+#if CARBON_POISON_VALUE_STORES
+    __asan_unpoison_memory_region(values_.data(),
+                                  values_.size() * sizeof(values_[0]));
+    all_poisoned_ = false;
+#endif
+  }
+  // Poison a single element.
+  auto PoisonElement([[maybe_unused]] int element) const -> void {
+#if CARBON_POISON_VALUE_STORES
+    __asan_unpoison_memory_region(values_.data() + element, sizeof(values_[0]));
+#endif
+  }
+  // Unpoison a single element.
+  auto UnpoisonElement([[maybe_unused]] int element) const -> void {
+#if CARBON_POISON_VALUE_STORES
+    __asan_unpoison_memory_region(values_.data() + element, sizeof(values_[0]));
+    all_poisoned_ = false;
+#endif
+  }
+
   // Set inline size to 0 because these will typically be too large for the
   // stack, while this does make File smaller.
   llvm::SmallVector<std::decay_t<ValueType>, 0> values_;
+
+#if CARBON_POISON_VALUE_STORES
+  // Whether the vector is currently fully poisoned.
+  mutable bool all_poisoned_ = true;
```

Comment:
Seems useful to capture the underlying reason for this optimization being so important. Maybe something like:

```suggestion
  // Whether the vector is currently fully poisoned.
  //
  // We use this to avoid repeated re-poisoning of the entire store. Doing so is linear in the size of the store, and we trigger re-poisoning frequently, for example on each import. Tracking that here allows us to coalesce these into a single linear operation.
  mutable bool all_poisoned_ = true;
```